### PR TITLE
[IMPROVEMENT] Added waiting time

### LINKED ISSUE #17: Display estimated time indication for queue
Clone of canihavesomecoffee/ccx_gitbot#4

Currently the test page (for a given id) just displays

```This test is still in the queue! Please check again later.```

This should be improved to show how many other tests are in the queue before you, and it should also show an estimate of how long that'll take before the test gets started (we have start + end time available of previous tests)

Expected: a PEP-8 compliant PR.

### PR DESCRIPTION
**Pull request by satyammittal**
**In raising this pull request, I confirm the following (please check boxes):**

- [X] I have read and understood the [contributors guide](https://github.com/canihavesomecoffee/sample-platform/blob/master/.github/CONTRIBUTING.md).
- [X] I have checked that another pull request for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**My familiarity with the project is as follows (check one):**

- [ ] I have never used the project.
- [ ] I have used the project briefly.
- [ ] I have used the project extensively, but have not contributed previously.
- [X] I am an active contributor to the project.

---

closes #17 
1) Added table telling issue number,repo and estimated waiting time
2) Time taken by current test in minutes

### CODE REVIEW

File: mod_test/controllers.py [hunk 1]
```diff
@@ -47,7 +47,10 @@ def index():
         'TestType': TestType
     }
 
-
+def diffdates(d1, d2):
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: Unused seemingly, so not needed in here.

File: mod_test/controllers.py [hunk 2]
```diff
@@ -56,6 +59,12 @@ def get_data_for_test(test, title=None):
         regressionTestCategoryLinkTable.c.category_id).subquery()
     categories = Category.query.filter(Category.id.in_(
         populated_categories)).order_by(Category.name.asc()).all()
+    comming_test=Test.query.filter(Test.progress == None).all()
+    last_running_test=Test.query.filter(Test.progress != None).order_by(Test.id.desc()).first()
+    if(last_running_test!=None):
+    pr=last_running_test.progress_data()
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: No indendation. Python won't like this.

File: templates/test/by_id.html [hunk 3]
```diff
@@ -46,7 +46,7 @@
                     {% endfor %}
                 </ol>
                 <br class="clear" />
-                {% if test.progress|length > 0 %}
+                {% if test.progress|length >  0 %}
```
Comment (canihavesomecoffee) [inline templates/test/by_id.html]: Not needed

File: templates/test/by_id.html [hunk 4]
```diff
@@ -67,7 +67,34 @@
                         </tbody>
                     </table>
                 {% else %}
-                    <p>This test is still in the queue! Please check again later.</p>
+                      <p>This test is still in the queue! Please check again later.Wait {{time}} minutes</p>
```
Comment (canihavesomecoffee) [inline templates/test/by_id.html]: {{ time }} is not containing the estimated time by when the test will be, so it's of no use to the user at all.

File: templates/test/by_id.html [hunk 5]
```diff
@@ -67,7 +67,34 @@
...
+                      <p>Other Tests in queue</p>
```
Comment (canihavesomecoffee) [inline templates/test/by_id.html]: There's no need to show the user this list.... It's already available on the index page. Just letting the user know there are still x tests queued before this one is sufficient.

**Comment by satyammittal** [comment]
OK I will fix that

File: templates/test/by_id.html (hunk 4 above)
Comment (satyammittal) [inline templates/test/by_id.html]: i have calculated time by taking the last active test and equating 
time=(test['end']-test['start']).totalMintutes
as the progress (start and end time) of tests in queue is empty.

Comment (satyammittal) [inline templates/test/by_id.html]: or this can be done by currentTest['start']-runningTest['end']

Comment (satyammittal) [inline templates/test/by_id.html]: ok i will  calculate the average on the last x tests and use that to predict

File: templates/test/by_id.html [hunk 6]
```diff
@@ -67,7 +67,7 @@
                         </tbody>
                     </table>
                 {% else %}
-                    <p>This test is still in the queue! Please check again later.</p>
+                      <p>This test is still in the queue! Please check again later.This test is still queued. There are {{ next }} tests queued before this one. Based on the average runtime, this test will be finished in {{ hr }} hours {{ min }} minutes.</p>
```
Comment (canihavesomecoffee) [inline templates/test/by_id.html]: Duplicate sentences

File: mod_test/controllers.py [hunk 7]
```diff
@@ -56,6 +55,36 @@ def get_data_for_test(test, title=None):
...
+    if len(test.progress) == 0:
+        comming_test = Test.query.filter(and_(Test.progress == None, Test.id < test.id)).all()
+        progress_test = Test.query.filter(and_(Test.progress != None, Test.id < test.id)).all()
+        total = 0
+        last_running_test_id_all = TestProgress.query.filter(TestProgress.status == TestStatus.completed).all()
+        canceled=TestProgress.query.filter(TestProgress.status == TestStatus.canceled).all()
+        running_test_before_this = len(comming_test) + len(progress_test) - len(last_running_test_id_all) - len(canceled)
+        #calculating average estimated time
+        if last_running_test_id_all != None:
+           for last_running_test_id in last_running_test_id_all:
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: Recalculate every time is a big no-no

File: mod_test/controllers.py [hunk 8]
```diff
@@ -56,6 +55,36 @@ def get_data_for_test(test, title=None):
...
+               last_running_test_id = last_running_test_id.test_id
+               last_running_test = Test.query.filter(Test.id == last_running_test_id).first()
+               pr = last_running_test.progress_data()
+               last_running_test = pr['end'] - pr['start']
+               last_running_test = last_running_test.total_seconds()
+               total += last_running_test
+            average_time = total // len(last_running_test_id_all)
+        time_run = 0
+        for pr_test in progress_test:
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: Too much overhead.

File: mod_test/controllers.py [hunk 9]
```diff
@@ -111,7 +129,11 @@ def get_data_for_test(test, title=None):
         'test': test,
         'TestType': TestType,
         'results': results,
-        'title': title
+        'title': title,
+        'next': running_test_before_this ,
+        'min': minutes,
+        'hr' : hours,
+        'gnew': u1
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: What is this doing in here?

File: mod_test/controllers.py [hunk 10]
```diff
@@ -56,6 +56,24 @@ def get_data_for_test(test, title=None):
         regressionTestCategoryLinkTable.c.category_id).subquery()
     categories = Category.query.filter(Category.id.in_(
         populated_categories)).order_by(Category.name.asc()).all()
+    hours=0
+    minutes=0
+    running_test_before_this="" 
+    #evaluating estimated time if the test is still in queue
+    if len(test.progress) == 0:
+        kvm_test = Kvm.query.filter(Kvm.test_id < test.id).first()
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: This entire section still can be done with better performance.

File: mod_test/controllers.py [hunk 11]
```diff
@@ -47,7 +48,6 @@ def index():
         'TestType': TestType
     }
 
-
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: Line must not be removed for PEP-8 compliancy. I urge you to edit the files with a full blown IDE (e.g. PyCharm, they have a free community edition)

File: mod_ci/controllers.py [hunk 12]
```diff
@@ -460,7 +460,33 @@ def progress_reporter(test_id, token):
...
+                        if last_running_test_id_all != None:
+                           for last_running_test_id in last_running_test_id_all:
+                               last_running_test_id = last_running_test_id.test_id
+                               last_running_test = Test.query.filter(Test.id == last_running_test_id).first()
+                               pr = last_running_test.progress_data()
+                               last_running_test = pr['end'] - pr['start']
+                               last_running_test = last_running_test.total_seconds()
+                               total += last_running_test
+                           average_time = total // len(last_running_test_id_all)
+                        newf = GeneralData('average_time',average_time)
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: This is most likely going to cause a multiple entries, messing up future calls.

File: mod_test/controllers.py (hunk 11 above)
Comment (satyammittal) [inline mod_test/controllers.py]: ok

**Comment by canihavesomecoffee** [comment]
Last commit still doesn't fix the issue with GeneralData....

**Comment by satyammittal** [comment]
I am working on it.. still finding the soln :)

File: mod_ci/controllers.py [hunk 13]
```diff
@@ -460,7 +460,38 @@ def progress_reporter(test_id, token):
                 repository = gh.repos(g.github['repository_owner'])(
                     g.github['repository'])
                 # If status is complete, remove the Kvm entry
+                u1 = GeneralData.query.filter(
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: You moved the last_running_test_id_all down, why didn't you move these two down into the if too?

File: mod_ci/controllers.py [hunk 14]
```diff
@@ -460,7 +460,38 @@ def progress_reporter(test_id, token):
...
+                    GeneralData.key == 'average_time').first()
+                average_time = 0
                 if status in [TestStatus.completed, TestStatus.canceled]:
+                    if u1 == None:
+                        last_running_test_id_all = TestProgress.query.filter(
+                            TestProgress.status == TestStatus.completed).all()    
+                        if last_running_test_id_all != None:
+                            for last_running_test_id in last_running_test_id_all:
+                                last_running_test_id = last_running_test_id.test_id
+                                last_running_test = Test.query.filter(
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: This is inefficient.

File: mod_ci/controllers.py [hunk 15]
```diff
@@ -460,7 +460,38 @@ def progress_reporter(test_id, token):
...
+                                total += last_running_test
+                            average_time = total // len(
+                                last_running_test_id_all)
+                            newf = GeneralData('average_time', average_time)
+                            g.db.add(newf)
+                            g.db.commit()
+                            average_time = float(newf.value)
+                    else:
+                        last_running_test_id_all = TestResult.query.count()
+                        number = len(last_running_test_id_all)
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: Won't work

File: mod_test/controllers.py [hunk 16]
```diff
@@ -56,19 +58,39 @@ def get_data_for_test(test, title=None):
...
+    hours = 0
+    minutes = 0
+    running_test_before_this = ""
+    # evaluating estimated time if the test is still in queue
+    if len(test.progress) == 0:
+        kvm_test = Kvm.query.filter(Kvm.test_id < test.id).first()
+        u1 = GeneralData.query.filter(
+            GeneralData.key == 'average_time').first()
+        average_time = float(u1.value)
+        running_test_before_this = len(kvm_test)
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: This is always 1

File: mod_ci/controllers.py [hunk 17]
```diff
@@ -461,6 +461,33 @@ def progress_reporter(test_id, token):
                     g.github['repository'])
                 # If status is complete, remove the Kvm entry
                 if status in [TestStatus.completed, TestStatus.canceled]:
+                    u1 = GeneralData.query.filter(
+                        GeneralData.key == 'average_time').first()
+                    average_time = 0
+                    if u1 == None:
+                        last_running_test_id_all = TestResult.all()    
+                        if last_running_test_id_all != None:
+                            for last_running_test_id in last_running_test_id_all:
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: This entire loop is still not very efficient. You fetch all test results, then select the test of that (through relationship), then fetch for each test all the progress data. If you go back to SQL, how would you make that a bit more efficient?

File: mod_test/controllers.py [hunk 18]
```diff
@@ -56,19 +58,39 @@ def get_data_for_test(test, title=None):
         regressionTestCategoryLinkTable.c.category_id).subquery()
     categories = Category.query.filter(Category.id.in_(
         populated_categories)).order_by(Category.name.asc()).all()
+    hours = 0
+    minutes = 0
+    running_test_before_this = ""
+    # evaluating estimated time if the test is still in queue
+    if len(test.progress) == 0:
+        kvm_test = Kvm.query.filter(Kvm.test_id < test.id)
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: Won't work, query isn't executed.

File: mod_test/controllers.py [hunk 19]
```diff
@@ -56,19 +58,39 @@ def get_data_for_test(test, title=None):
...
+        u1 = GeneralData.query.filter(
+            GeneralData.key == 'average_time').first()
+        average_time = float(u1.value)
+        running_test_before_this = len(kvm_test)
+        for pr_test in kvm_test:
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: Still not very efficient... Same remark as for the other loop.

File: mod_ci/controllers.py [hunk 20]
```diff
@@ -461,6 +464,44 @@ def progress_reporter(test_id, token):
...
+                            average_time += sec
+                        if len(finished_tests) !=0 :
+                            average_time = total // len(
+                                finished_tests)
+                        newf = GeneralData('average_time', average_time)
+                        g.db.add(newf)
+                        g.db.commit()
+                        average_time = float(newf.value)
+                    else:
+                        number = TestResult.query.count()
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: Incorrect. `TestResult` table contains information about individual regression test and will give count = no. of tests * no. of regression tests.

File: mod_ci/controllers.py [hunk 21]
```diff
@@ -461,6 +464,44 @@ def progress_reporter(test_id, token):
...
+                        for p in times:
+                            k = p.time.split(',')
+                            pr1 = datetime.strptime(k[0], '%Y-%m-%d %H:%M:%S')
+                            pr2 = datetime.strptime(k[1], '%Y-%m-%d %H:%M:%S')
+                            sec = (pr1-pr2).total_seconds()
+                            if sec < 0 :
+                                sec *= -1
+                            average_time += sec
+                        if len(finished_tests) !=0 :
+                            average_time = total // len(
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: Where is total coming from?

File: mod_ci/controllers.py [hunk 22]
```diff
@@ -461,6 +464,44 @@ def progress_reporter(test_id, token):
...
+                                        [TestStatus.building,TestStatus.completed,TestStatus.canceled]))
+                                    ).subquery()
+                        finished_tests = times2
+                        times = g.db.query(finished_tests.c.test_id,label('time',func.group_concat(
+                                finished_tests.c.timestamp))).group_by(finished_tests.c.test_id).all()
+                        for p in times:
+                            k = p.time.split(',')
+                            pr1 = datetime.strptime(k[0], '%Y-%m-%d %H:%M:%S')
+                            pr2 = datetime.strptime(k[1], '%Y-%m-%d %H:%M:%S')
+                            sec = (pr1-pr2).total_seconds()
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: We can safely assume that pr2 always will be later than pr1. No need to check for negative values

File: mod_ci/controllers.py [hunk 23]
```diff
@@ -461,6 +464,44 @@ def progress_reporter(test_id, token):
...
                 # If status is complete, remove the Kvm entry
                 if status in [TestStatus.completed, TestStatus.canceled]:
+                    u1 = GeneralData.query.filter(
+                        GeneralData.key == 'average_time').first()
+                    average_time = 0
+                    if u1 == None:
+                        finished_tests = g.db.query(TestProgress.test_id).filter(
+                                        TestProgress.status.in_([TestStatus.canceled, TestStatus.completed])
+                                        ).subquery()
+                        times2 = g.db.query(TestProgress).filter(
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: You're not taking the preparation state into account.

File: mod_ci/controllers.py [hunk 24]
```diff
@@ -461,6 +464,44 @@ def progress_reporter(test_id, token):
...
+                                    and_(TestProgress.test_id.in_(finished_tests), TestProgress.status.in_(
+                                        [TestStatus.building,TestStatus.completed,TestStatus.canceled]))
+                                    ).subquery()
+                        finished_tests = times2
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: Reassigning values is getting messy and unreadable

File: mod_ci/controllers.py [hunk 25]
```diff
@@ -461,6 +464,44 @@ def progress_reporter(test_id, token):
                     g.github['repository'])
                 # If status is complete, remove the Kvm entry
                 if status in [TestStatus.completed, TestStatus.canceled]:
+                    u1 = GeneralData.query.filter(
+                        GeneralData.key == 'average_time').first()
+                    average_time = 0
+                    if u1 == None:
```
Comment (canihavesomecoffee) [inline mod_ci/controllers.py]: if u1 is None is preferred over u1 == None

File: mod_ci/controllers.py (hunk 20 above)
Comment (satyammittal) [inline mod_ci/controllers.py]: Ok thanks for pointing out. i will look into it

File: mod_ci/controllers.py (hunk 22 above)
Comment (satyammittal) [inline mod_ci/controllers.py]: done :)

File: mod_test/controllers.py [hunk 26]
```diff
@@ -56,19 +60,47 @@ def get_data_for_test(test, title=None):
...
+    minutes = 0
+    running_test_before_this = ""
+    # evaluating estimated time if the test is still in queue
+    if len(test.progress) == 0:
+        
+        kvm_test = g.db.query(Kvm.test_id).filter(Kvm.test_id < test.id).subquery()
+        times2 = g.db.query(TestProgress.test_id,label('time',func.group_concat(TestProgress.timestamp))).filter(
+                TestProgress.test_id.in_(kvm_test)
+                ).group_by(TestProgress.test_id).all()
+        kvm_test = times2
```
Comment (canihavesomecoffee) [inline mod_test/controllers.py]: Why re-using the variables? Doesn't help readability

File: templates/test/by_id.html [hunk 27]
```diff
@@ -7,6 +7,7 @@
     <div class="row">
         <div class="row">
             <div class="column medium-8">
+            {{gnew}}
```
Comment (canihavesomecoffee) [inline templates/test/by_id.html]: Please remove

**Pull request merged**